tweet_recommendations/other_methods/dbscan_based_method.py

```diff
@@ -0,0 +1,94 @@
+from collections import namedtuple
+from itertools import chain
+from typing import *
+
+import numpy as np
+import pandas as pd
+from scipy.spatial.distance import cosine
+from sklearn.cluster import DBSCAN
+from sklearn.neighbors.kd_tree import KDTree
+
+from ..estimator import Estimator
+
+CentroidData = namedtuple("CentroidData", ["cluster_medoid", "popular_hashtags"])
+
+
+class DBScanBasedEstimator(Estimator):
+    def __init__(self,
+                 verbose: bool = False):
+        self._clusters: np.ndarray = None
+
+        self.clusterizer = None
+        self.neighbours = None
+        self.centroids_data = {}
+
+        self.verbose = verbose
+
+    def fit(self, x: pd.DataFrame, y: Optional[pd.DataFrame] = None, **fit_params):
+        """
+        Builds hashtags representations using w2v method and clusters them using DBScan.
+        :param x: pd.DataFrame with tweet content, user id, and separate hashtags.
+                It is "original_tweets.p" in our case.
+        :param y: None, needed for compatibility.
+        :param fit_params:
+            minimal_hashtag_occurrence: int. If hashtag occurred less than this number
+            then it's not considered during prediction (simply removed).
+            To include all hashtags put number <= 0.
+        :return: self.
+        """
+        minimal_hashtag_occurence = fit_params["minimal_hashtag_occurence"]
+
+        x = self.drop_tweets_with_hashtags_that_occurred_less_than(x, minimal_hashtag_occurence)
+        embeddings = np.array(x["embedding"].to_list())
```

Comment:
Z @jmichalowski12 głowiliśmy się trochę nad kwestią `pd.Series`, które zawierają pojedyncze `np.ndarray` – ktoś ma jakiś sposób jak łatwo przejść z takiego formatu na macierz?

Reply:
Sam nie znam sposobu, na stacku w jednym z wątków też ktoś zamienia na listę

Reply:
Sam nie znam takiego sposobu, a na stacku w jednym z wątków też ktoś zamieniał najpierw na listę.

Reply:
A nie wystarczy jakiś np.vstack czy coś podobnego?
`np.vstack(df['column'].values)`
kojarzy mi się, że to powinno być jedno z szybszych podejść

Reply:
Sprawdziłem. Dla `x` wierszy w tabeli, każdy o `n` słowach, gdzie każde słowo ma wektor `m`, `vstack` powoduje powstanie macierzy 2-wymiarowej: 
![image](https://user-images.githubusercontent.com/16157910/55668914-afc4c300-5870-11e9-8105-acb3aec720e0.png),
natomiast samo `stack` macierzy 3-wymiarowej:
![image](https://user-images.githubusercontent.com/16157910/55668921-beab7580-5870-11e9-8f0b-3facf7569317.png)